component/file_cache/file_cache_async.go

```diff
@@ -0,0 +1,331 @@
+				time.Sleep(time.Duration(restTime) * (time.Millisecond))
+				val, _ := fc.fileOps.Load(key)
+
+				fileOperation := val.(FileAttributes).operation
+				fileOptions := val.(FileAttributes).options
+
+				if val != nil {
+					log.Trace("AsyncFileCache:: async_cloud_handler : The key in the function call is %s and the value is %s", key, fileOptions)
+					switch {
+					case fileOperation == "DeleteDir":
+						returnVal = fc.asyncDeleteDir(fileOptions.(internal.DeleteDirOptions))
+
+					case fileOperation == "RenameDir":
+						returnVal = fc.asyncRenameDir(fileOptions.(internal.RenameDirOptions))
+
+					case fileOperation == "CreateFile":
+						returnVal = fc.asyncCreateFile(fileOptions.(internal.CreateFileOptions))
+						// fc.fileOps.Delete(key)
+
+					case fileOperation == "DeleteFile":
+						returnVal = fc.asyncDeleteFile(fileOptions.(internal.DeleteFileOptions))
+
+					case fileOperation == "FlushFile":
+						returnVal = fc.asyncFlushFile(fileOptions.(FlushFileAbstraction))
+
+					case fileOperation == "RenameFile":
+						returnVal = fc.asyncRenameFile(fileOptions.(internal.RenameFileOptions))
+
+					case fileOperation == "CreateDir":
+						returnVal = fc.asyncCreateDir(fileOptions.(internal.CreateDirOptions))
+
+					case fileOperation == "Chmod":
+						returnVal = fc.asyncChmod(fileOptions.(internal.ChmodOptions))
+
+					case fileOperation == "Chown":
+						returnVal = fc.asyncChown(fileOptions.(internal.ChownOptions))
+
+					case fileOperation == "SyncFile":
+						returnVal = fc.asyncSyncFile(fileOptions.(internal.SyncFileOptions))
+
+					case fileOperation == "ChmodAndFlush":
+						returnVal = fc.asyncChmodAndFlush(fileOptions.(internal.ChmodOptions))
+					}
+
+					log.Trace("AsyncFileCache:: async_cloud_handler: The key after the function call is %s and the value is %s", key, fileOptions)
+
+					if returnVal == nil {
+						log.Trace("AsyncFileCache:: async_cloud_handler: File name %s has just finished file operation %s", key, fileOperation)
+						tries = 0                                 //attempt was successful, reset try counter
+						_ = fc.fileOps.CompareAndDelete(key, val) //file has been serviced, remove it from map only if file op hasn't been updated
+
+					} else {
+
+						numFailed++
+						if tries < maxTries {
+
+							tries++ //failed op, increase timeout duration
+						}
+
+					}
+				}
+				return true
+			})
+		}
+
+		//check channel to end thread here
+		// if <-fc.closeSignal == 1 {
+		// 	return
+		// }
+	}
+
+}
+
+// File is already flushed locally, we just need to upload it to the cloud
+func (fc *FileCache) asyncFlushFile(options FlushFileAbstraction) error {
+
+	localPath := common.JoinUnixFilepath(fc.tmpPath, options.Name)
+	uploadHandle, err := common.Open(localPath)
+
+	if err != nil {
+		log.Err("FileCache::FlushFile : error [unable to open upload handle] %s [%s]", options.Name, err.Error())
+		return err
+	}
+
+	err = fc.NextComponent().CopyFromFile(
```

Comment:
Should this also call the next components FlushFile?

Reply:
In the original design, fileCache flushFile used copyFromFile so I just did the same thing here